Mathlib/AlgebraicTopology/SimplicialObject.lean

```diff
@@ -0,0 +1,874 @@
+/-
+Copyright (c) 2021 Scott Morrison. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Johan Commelin, Scott Morrison, Adam Topaz
+
+! This file was ported from Lean 3 source module algebraic_topology.simplicial_object
+! leanprover-community/mathlib commit 5ed51dc37c6b891b79314ee11a50adc2b1df6fd6
+! Please do not edit these lines, except to modify the commit id
+! if you have ported upstream changes.
+-/
+import Mathlib.AlgebraicTopology.SimplexCategory
+import Mathlib.CategoryTheory.Arrow
+import Mathlib.CategoryTheory.Limits.FunctorCategory
+import Mathlib.CategoryTheory.Opposites
+
+/-!
+# Simplicial objects in a category.
+
+A simplicial object in a category `C` is a `C`-valued presheaf on `SimplexCategory`.
+(Similarly a cosimplicial object is functor `SimplexCategory ⥤ C`.)
+
+Use the notation `X _[n]` in the `Simplicial` locale to obtain the `n`-th term of a
+(co)simplicial object `X`, where `n` is a natural number.
+
+-/
+
+
+open Opposite
+
+open CategoryTheory
+
+open CategoryTheory.Limits
+
+universe v u v' u'
+
+namespace CategoryTheory
+
+variable (C : Type u) [Category.{v} C]
+
+-- porting note: removed @[nolint has_nonempty_instance]
+/-- The category of simplicial objects valued in a category `C`.
+This is the category of contravariant functors from `SimplexCategory` to `C`. -/
+def SimplicialObject :=
+  SimplexCategoryᵒᵖ ⥤ C
+#align category_theory.simplicial_object CategoryTheory.SimplicialObject
+
+@[simps!]
```

Comment:
I'm a bit confused about all these `simps!` attributes. There was nothing similar in mathlib3 right? Are they strictly necessary, and would it perhaps be better to make `SimplicialObject` a `reducible` instead?

Reply:
Without the `reducible` attribute, the `simps!` for the Category instance (and extra `ext` lemmas for morphisms) were necessary. I have made `SimplicialObject` reducible in the last commit.

Reply:
Okay, I'll ping @jcommelin , @semorrison and @adamtopaz to see if they're happy with the change.

Reply:
I mean, I'd prefer that we don't make it reducible. It's more work to set up here, unfortunately. We have a tendency to regret making things less opaque, however...

Reply:
Ok. I have reverted back to the previous situation where `SimplicialObject` is not declared reducible.

Reply:
I'll let someone else take over the review of this PR, but I'm not sure that `simp` lemmas unfolding the definition of a type is a particularly good idea either for the same reasons. Probably a lot of the simp lemmas that become accessible after unfolding definitions should be proved on `SimplicialObject` directly.

Reply:
Indeed, a lemma like
```lean
theorem SimplicialObject.δ_naturality {X' X : SimplicialObject C} (f : X ⟶ X') {n : ℕ} (i : Fin (n + 2)) :
    X.δ i ≫ f.app (op [n]) = f.app (op [n + 1]) ≫ X'.δ i```
is a particular case of `NatTrans.naturality`, and it is stated as a `simp` lemma that is specific to `SimplicialObject`.

Reply:
Sorry, a bit of a tangle here. I have a PR open on mathlib proposing to adjust the simps projections for `category` so we don't unfold `Hom`. I should have got that done and forwarded port faster...

Reply:
Okay, I propose that I will quickly forward port that PR, and then merge that into this before we proceed?